go/mysql/auth_server.go

```diff
@@ -17,109 +17,233 @@ limitations under the License.
+	// methods during the initial handshake.
+	//
+	// It's not needed to also support those methods in the AuthMethods(),
+	// in fact, if you want to only support for example clear text passwords,
+	// you must still return `mysql_native_password` or `caching_sha2_password`
+	// here and the auth switch protocol will be used to switch to clear text.
+	DefaultAuthMethodDescription() AuthMethodDescription
+}
+
+// AuthMethod interface for concrete auth method implementations.
+// When building an auth server, you usually don't implement these yourself
+// but the helper methods to build AuthMethod instances should be used.
+type AuthMethod interface {
+	// Name returns the auth method description for this implementation.
+	// This is the name that is sent as the auth plugin name during the
+	// Mysql authentication protocol handshake.
+	Name() AuthMethodDescription
+
+	// HandleUser verifies if the current auth method can authenticate
+	// the given user with the current auth method. This can be useful
+	// for example if you only have a plain text of hashed password
+	// for specific users and not all users and auth method support
+	// depends on what you have.
+	HandleUser(conn *Conn, user string) bool
+
+	// AuthPluginData generates the information for the auth plugin.
+	// This is included in for example the auth switch request. This
+	// is auth plugin specific and opaque to the Mysql handshake
+	// protocol.
+	AuthPluginData() ([]byte, error)
+
+	// HandleAuthPluginData handles the returned auth plugin data from
+	// the client. The original data the server sent is also included
+	// which can include things like the salt for `mysql_native_password`.
+	//
+	// The remote address is provided for plugins that also want to
+	// do additional checks like IP based restrictions.
+	HandleAuthPluginData(conn *Conn, user string, serverAuthPluginData []byte, clientAuthPluginData []byte, remoteAddr net.Addr) (Getter, error)
+}
+
+// UserValidator is an interface that allows checking if a specific
+// user will work for an auth method. This interface is called by
+// all the default helpers that create AuthMethod instances for
+// the various supported Mysql authentication methods.
+type UserValidator interface {
+	HandleUser(user string) bool
+}
+
+// CacheState is a state that is returned by the UserEntryWithCacheHash
+// method from the CachingStorage interface. This state is needed to indicate
+// whether the authentication is accepted, rejected by the cache itself
+// or if the cache can't fullfill the request. In that case it indicates
+// that with AuthNeedMoreData.
+type CacheState int
+
+const (
+	// AuthRejected is used when the cache knows the request can be rejected.
+	AuthRejected CacheState = iota
+	// AuthAccepted is used when the cache knows the request can be accepted.
+	AuthAccepted
+	// AuthNeedMoreData is used when the cache doesn't know the answer and more data is needed.
+	AuthNeedMoreData
```

Comment:
not able to find the usages of `AuthNeedMoreData`

Reply:
@harshit-gangal Ah yeah, that's because there's 3 possible states and the other two get checked and the fallback then equates this one. I can also change this to a switch statement if that's preferred with a default that returns an error on an unknown state?

Reply:
I've updated it to a switch statement to be more explicit here and reference the value.